fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: rtx.messages,
+            log,
+            data: None,
+        };
+        Ok(r)
+    } else {
+        let swap_fee = amount;
+        let effective_amount = Uint128::zero();
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: vec![],
+            log,
+            data: None,
+        };
+        Ok(r)
+    }
+}
+
+/* ***************************************************
+ * *****************    Helpers      *****************
+ * ***************************************************/
+
+fn amount_from_funds(funds: &Vec<Coin>) -> Uint128 {
+    // TODO(LR) return proper error
+    if funds.len() == 1
+    /* && funds[0].denom == "fet" */
+    {
+        // TODO(LR) does cosmwas allows sending multiple funds of the same token
+        funds[0].amount
+    } else {
+        Uint128::zero()
+    }
+}
+
+fn send_tokens<A: Api>(
+    api: &A,
+    env: &Env,
+    state: &State,
+    to_address: &CanonicalAddr,
+    amount: Uint128,
+    action: &str,
+) -> HandleResult {
+    let from_address = &env.contract.address;
+    let from_human = api.human_address(&from_address)?;
+    let to_human = api.human_address(to_address)?;
+    let log = vec![log("action", action), log("to", to_human.as_str())];
+    let coin = Coin {
+        amount,
+        denom: state.denom.clone(),
+    };
+
+    let r = HandleResponse {
+        messages: vec![CosmosMsg::Bank(BankMsg::Send {
+            from_address: from_human,
+            to_address: to_human,
+            amount: vec![coin],
+        })],
+        log,
+        data: None,
+    };
+    Ok(r)
+}
+
+/* ***************************************************
+ * ***************    Verifiers      *****************
+ * ***************************************************/
+
+fn verify_tx_relay_eon(eon: u64, state: &State) -> HandleResult {
+    if eon != state.relay_eon {
+        Err(StdError::generic_err(
+            "Tx doesn't belong to current relayEon",
+        ))
+    } else {
+        Ok(HandleResponse::default())
+    }
+}
+
+fn verify_not_paused(env: &Env, state: &State) -> HandleResult {
+    if env.block.height >= state.paused_since_block {
+        Err(StdError::generic_err(format!(
+            "Contract has been paused {}",
+            state.paused_since_block
+        )))
+    } else {
+        Ok(HandleResponse::default())
+    }
```

Comment:
Just nit picking: it would be better to use just `<` instead of `>=`:

```suggestion
    if env.block.height < state.paused_since_block {
        Ok(HandleResponse::default())
    } else {
        Err(StdError::generic_err(format!(
            "Contract has been paused {}",
            state.paused_since_block
        )))
    }
```